File: Data_Structures_Exercises_C#_Python/1_List/C#_List/ListManager.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;

namespace Data_Structures_Exercises
{
    public class ListManager
    {
        private int[] list;

        // Constructor to initialize the list
        public ListManager()
        {
            list = new int[0];
        }

        /// <summary>
        /// Method to add an element to the end of the list (case 1)
        /// </summary>  
        public void AddToEnd()
        {
            Console.WriteLine("Enter the number to add to the list: ");
            int newElement;

            // Validate user input
            while (!int.TryParse(Console.ReadLine(), out newElement))
            {
                Console.WriteLine("Invalid input. Please enter a valid number: ");
            }

            // Create a new list with an extra space
            int[] newList = new int[list.Length + 1];

            // Copy the elements from the old list to the new list
            for (int i = 0; i < list.Length; i++)
            {
                newList[i] = list[i];
            }

            // Add the new element to the end of the new list
            newList[list.Length] = newElement;

            // Update the reference to the new list
            list = newList;

            Console.WriteLine($"Element {newElement} added to the end of the list.");
        }

        /// <summary>
        /// Method to insert an element at a given location (case 2)
        /// </summary>
        public void InsertElement(int index, int newValue)
        {
            if (index < 0 || index > list.Length)
            {
                Console.WriteLine("Invalid index.");
                return;
            }
            // Create a new list with an extra space
            int[] newList = new int[list.Length + 1];

            // Copy the elements from the old list to the new list
            for (int i = 0; i < index; i++)
            {
                newList[i] = list[i];
            }
            // Insert the new element at the specified index
            newList[index] = newValue;

            // Copy the remaining elements from the old list to the new list
            for (int i = index; i < list.Length; i++)
            {
                newList[i + 1] = list[i];
            }
            // Update the reference to the new list
            list = newList;
            Console.WriteLine($"Inserted element {newValue} at index {index}.");
        }

        /// <summary>
        /// Method to delete the element at a given location (case 3)
        /// </summary>
        public void DeleteElement(int index)
        {
            if (index < 0 || index >= list.Length)
            {
                Console.WriteLine("Invalid index.");
                return;
            }
            // Create a new list with one less space
            int[] newList = new int[list.Length - 1];

            for (int i = 0; i < index; i++)
            {
                newList[i] = list[i];
            }
            // Copy elements after the deleted index
            for (int i = index + 1; i < list.Length; i++)
            {
                newList[i - 1] = list[i];
            }
            // Update the reference to the new list
            list = newList;

            Console.WriteLine($"Element at index {index} has been deleted.");

        }

        /// <summary>
        /// Method to display the list (case 4)
        /// </summary>
        public void DisplayList()
        {
            if (list.Length == 0)
            {
                Console.WriteLine("The list is empty.\n");
                return;
            }
            /*
            Console.WriteLine("These are the elements in your list.");
            for (int i = 0; i < list.Length; i++)
            {
                Console.WriteLine($"Index {i}: {list[i]}");
            }*/
            Console.WriteLine("List elements: [" + string.Join(", ", list) + "]\n");
        }

        /// <summary>
        /// Method to display the count of the total number of elements in the list (case 5)
        /// </summary>
        public void DisplayCount()
        {
            int count = list.Length;
            Console.WriteLine($"The list contains {count} elements.");
        }

        /// <summary>
        /// Method to clear (initialize) the list (case 6)
        /// </summary> 
        public void ClearList()
        {
            list = new int[0];
            Console.WriteLine("List has been cleared.");
        }

    }
}

File: Data_Structures_Exercises_C#_Python/1_List/Python_List/ListManager.py
from textwrap import dedent

class ListManager:
    def __init__(self):
        self.list = []

    def display_menu(self):
        print(dedent("""
        A menu of options:
        Press 1 to add an element to the end of the list.
        Press 2 to insert an element at a given location.
        Press 3 to delete an element from a given location.
        Press 4 to display all of the list elements.
        Press 5 to count the number of elements in the list.
        Press 6 to clear (initialize) the list.
        Press 7 to exit the program.
        """))

    def run(self):
        is_running = True
        while is_running:
            self.display_menu()
            choice = input("Enter your choice:\n")
            if choice == '1':
                self.add_to_end()
            elif choice == '2':
                self.insert_element()
            elif choice == '3':
                self.delete_element()
            elif choice == '4':
                self.display_elements()
            elif choice == '5':
                self.count_elements()
            elif choice == '6':
                self.clear_list()
            elif choice == '7':
                is_running = False
            else:
                print("Invalid choice. Please try again.")

    # Choice 1
    def add_to_end(self):
        while True:
            item = input("Enter an element to add to the end of the list.\n")
            if item: # Check if the input is not empty
                self.list.append(item) # Add item to the end of the list
                print(f"'{item}' added to the list.")
                print(f"The list is now: {self.list}")
                break
            else:
                print("Invalid input. Please try again.")

    # Choice 2
    def insert_element(self):
        while True:
            try:
                index = int(input("Enter the index where you want to insert the element.\n"))
                value = input("Enter the element to insert.\n")
                self.list.insert(index, value)
                print(f"'{value}' inserted at index {index}.")
                print(f"The list is now: {self.list}")
                break
            except ValueError:
                print("Invalid input. Please enter a valid index.")
            except IndexError:
                print("Invalid index. Please enter a valid index.")

    # Choice 3
    def delete_element(self):
        while True:
            try:
                index = int(input("Enter the index of the element to delete.\n"))
                value = self.list.pop(index)
                print(f"'{value}' deleted from index {index}.")
                print(f"The list is now: {self.list}")
                break
            except ValueError:
                print("Invalid input. Please enter a valid index.")
            except IndexError:
                print("Invalid index. Please enter a valid index.")

    # Choice 4
    def display_elements(self):
        if self.list:
            print(f"The list elements are: {self.list}")
            
        else:
            print("The list is empty.")

    # Choice 5
    def count_elements(self):
        print(f"The number of elements in the list is: {len(self.list)}")

    # Choice 6
    def clear_list(self):
        self.list = []
        print("The list has been cleared.")